espnet2/train/trainer.py

```diff
@@ -391,7 +381,26 @@ def train_one_epoch(
 
             with autocast(scaler is not None):
                 with reporter.measure_time("forward_time"):
-                    loss, stats, weight = model(**batch)
+                    retval = model(**batch)
+
+                    # Note(kamo):
+                    # Supporting two patterns for the returned value from the model
+                    #   a. dict type
+                    if isinstance(retval, dict):
+                        loss = retval["loss"]
+                        stats = retval["stats"]
+                        weight = retval["weight"]
+
+                        if "optim_idx" in retval:
+                            optim_idx = retval["optim_idx"]
+                        else:
+                            optim_idx = None
+
+                    #   b. tuple or list type
+                    else:
+                        loss, stats, weight = retval
+                        optim_idx = None
+
```

Comment:
I also added a system to select a optimizer to be used from the model for each step:

```python
class YourModel:
    def forward(self, ...):
        ...
        # The 0th optimizer will be called
        return {"loss": loss, "stats": stats, "weight": weight, "optim_idx": 0}
```